Add feature to attach all pdfs in given directory

phrb opened the pull request:
This is a feature suggestion. I implemented this so I could create an org-file
attaching my existing pdf files associated with my bib file.

Currently the script copies all pdfs to the `data` directory. An interesting option
could be only symlinking the files.

With this modification, the `.doirc` file can now contain a `pdfpath` argument specifying a path
where all pdf files in the `.bib` file can be found.

The script will then search for pdfs in the directory with the exact bib
entry key. For example, for an entry `authorDateTitle`, the script
expects to find `authorDateTitle.pdf` in the directory tree given by
`pdfpath`.

I also updated the file hashes for the tests, which seem to have changed again.

Another issue I am having during the tests is this:
```
DOI_to_org/doi_to_org.py:148: DeprecationWarning: plaintext() is deprecated since 0.19: use __unicode__() instead
  return name.plaintext()
```
I tried using `__unicode__` in line 148 but it did not work. Any idea where this is from? It does not break the tests though.

phrb committed:
Add feature to attach all pdfs in given directory

The `.doirc` file can now contain a `pdfpath` argument specifying a path
where all pdf files in the `.bib` file can be found.

The script will then search for pdfs in the directory with the exact bib
entry key. For example, for an entry `authorDateTitle`, the script
expects to find `authorDateTitle.pdf` in the directory tree given by
`pdfpath`.

phrb committed:
Update file hash

Ezibenroc commented on doi_to_org.py:
```diff
@@ -204,6 +209,20 @@ def tempfile_from_path(cls, path):
+                    shutil.copyfile(os.path.join(root, file), f.name)
+                    return f
+
+        sys.exit("Couldn't find file '%s.pdf' in '%s'" % (key, path))
```
I think it would be better to raise an exception here, like `FileNotFound`, and if we catch it later then calling `sys.exit`.

Ezibenroc replied:
We could easily allow another extension (or no extension at all) here. We just need to replace the conditoin from line 219 by a `os.path.splitext(file)[0] == key`.

Ezibenroc replied:
Maybe rename `file`, it is already a class in Python 2.

Ezibenroc commented on doi_to_org.py:
```diff
@@ -305,7 +329,10 @@ def orgmode_from_bibentry(self, attached_file_name=None, attached_file_hash=None
+        if self.pdfpath:
+            return self.bibentry.key + self.attachment.extension
+        else:
+            return self.bibentry.title.replace(' ', '_') + self.attachment.extension
```
I don't get this change. This filename will be the name of the attached file, it is not necessary the same as the original file. So why do you want to use another naming convention depending on the origin of the file?

Ezibenroc replied:
And because of this change, `OrgEntry` has to know `pdfpath`. I don't think this should be the case, we just need to give to the `OrgEntry` an `Attachment`.

phrb replied:
Thanks for your feedback!

I added this in order to keep the previous filenames I had but I agree this was not the best solution.

Do you think adding an option to the `.doirc` file to keep names or select different strategies for naming would be a better way to achieve this?

Ezibenroc replied:
Yes, we could add an option to the `.doirc` file.
An easy way would be a boolean variable that tells if we keep the original name or not. A more complex solution could be a "format string" that tells what to use, like this:
```yaml
pdfname : "{bibtexkey}_{authorname}_{title}"
```
or :
```yaml
pdfname : "{filename}"
```
I am not sure about the best syntax however, and it will be harder to do.

Ezibenroc commented:
That's a great idea of feature!

> Currently the script copies all pdfs to the data directory. An interesting option
could be only symlinking the files.

Maybe as an option. By default, I think it is best to copy the files. A third nice option would be to move them.

> The script will then search for pdfs in the directory with the exact bib entry key. For example, for an entry authorDateTitle, the script expects to find authorDateTitle.pdf in the directory tree given by pdfpath

Maybe we could relax this constraint and take the file authoreDateTitle.ext, where "ext" can be anything (pdf, ps, zip, txt...). We could also allow a file without extension.
What do you think?

> I also updated the file hashes for the tests, which seem to have changed again.

Argh...

> I tried using __unicode__ in line 148 but it did not work. Any idea where this is from? It does not break the tests though.

Yes, I also had this problem, I did not found how to fix the warning (the thing with __unicode__ did not work).

Ezibenroc commented:
I see that you added a pdf in the test data. I am not sure it is a good idea, for two reasons:
- The file is not super large, but still. This is a lot of space only for test purpose...
- I am not sure we have (legally) the right to have it publicly available here, even though it is on HAL.

Maybe use instead a small text file?

phrb committed:
Apply changes requested in the PR review

- Raise `FileNotFoundException` instead of exiting
- Accept other file extensions
- Rename `file` variable
- Revert to single naming strategy
- Add a blank and small pdf (~11kb) for testing
- Update file hashes for testing

phrb commented:
Ok, it took me long enough but here are the changes!

Ezibenroc commented on doi_to_org.py:
```diff
@@ -378,5 +374,10 @@ def org_entry_fabric(orgfile, pdfpath, arg):
+        except FileNotFoundError as e:
+            print(e)
+            sys.exit()
+
```
This would be better:
```python
sys.exit(e)
```
It prints the error message on stderr instead of stdout and it is one line less.

Ezibenroc commented on doi_to_org.py:
```diff
@@ -204,6 +209,20 @@ def tempfile_from_path(cls, path):
+                    return temp_f
+
+        raise FileNotFoundError("Couldn't find file '%s.pdf' in '%s'" % (key, path))
+
```
This is not always `key.pdf` now, as we relaxed this constraint. Maybe this message could do:
```python
raise FileNotFoundError("Could not find a file starting by '%s' in '%s'" % (key, path))
```

Ezibenroc commented:
Just some minor changes and we are good to go :+1:

phrb commented:
Thanks for the feedback Tom, I'll fix it soon!

phrb committed:
Improve error handling

phrb commented:
Done!

Ezibenroc commented:
Thanks!
